pp/file/file_rpc.go

```diff
@@ -0,0 +1,117 @@
+package file
+
+import (
+	"strconv"
+	"strings"
+	"time"
+	"sync"
+	"io"
+
+	"github.com/stratosnet/sds/msg/protos"
+)
+
+var (
+	reFileMutex sync.Mutex
+	// key(fileHash) : value(fileSize)
+	rpcFileInfoMap = make(map[string]uint64)
+	// key(fileHash) : value(event)
+	reFileEvent = make(map[string][]string)
+	// key(fileHash) : value(pipe)
+	pipes = make(map[string]pipe)
+)
+
+type pipe struct {
+	reader   *io.PipeReader
+	writer   *io.PipeWriter
+}
+
+// IsFileRpcRemote
+func IsFileRpcRemote(hash string) bool {
+	str := fileMap[hash]
+	if str == "" {
+		return false
+	}
+	return strings.Split(str, ":")[0] == "rpc"
+}
+
+// GetRemoteFileData
+func GetRemoteFileData(hash string, offset *protos.SliceOffset) []byte {
+    // compose event, as well notify the remote user
+	var r []string
+	r = append(r, "1") //return code
+	r = append(r, strconv.FormatUint(offset.SliceOffsetStart, 10)) // extended result: slice start
+	r = append(r, strconv.FormatUint(offset.SliceOffsetEnd, 10)) // extended result: slice end
+
+	// send event and open the pipe for coming data
+    reFileMutex.Lock()
+	reFileEvent[hash] = r
+	var p pipe
+	p.reader, p.writer = io.Pipe()
+	pipes[hash] = p
+    reFileMutex.Unlock()
+
+	// read on pipe
+	data := make([]byte, offset.SliceOffsetEnd - offset.SliceOffsetStart)
+	var cursor []byte
+	var read uint64
+	cursor = data[:]
+	select {
+	case <-time.After(3 * time.Second):
```

Comment:
make the wait time a const so it can be easier captured from code

Reply:
Yes. Modified. Actually, I realized that the timeout didn't work. Now it works.